Refactor WidgetPod::event.

xStrom opened the pull request:
This PR is just the refactoring part of #986 so that we can get the refactoring in first and make the command targeting work more clear.

This new approach is more like `WidgetPod::lifecycle`. It doesn't copy/clone as much and is fewer lines of code allowing for easier understanding.

cmyr commented on druid/src/core.rs:
```diff
@@ -558,159 +561,168 @@ impl<T: Data, W: Widget<T>> WidgetPod<T, W> {
+        // If we need to replace either the event or its data.
+        let mut modified_event = None;
+        // Tracking whether we need to merge our state up to the parent.
+        let mut needs_merge = false;
```
is this an optimization (my assumption), or are there times when merging would be incorrect?

if an optimization: my gut feeling is that merging is only a trivial amount of work in the general case, such that this optimization is unlikely to be meaningful even over very large graphs, and I'm not sure it's something it is worth adding more logic to handle, at least without being guided by a concrete performance concern?

xStrom commented:
I don' think the `needs_merge` adds any significant complexity, but I also don't think it's a significant performance win right now. So I removed it for clarity and added a comment about this situation.
